std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+    }
+
+    // L/A initialiser
+    this(ComponentType l, ComponentType a = 0)
+    {
+        foreach(c; TypeTuple!('l','r','g','b'))
+            mixin(ComponentExpression!("this._ = l;", c, null));
+        static if(CanFind!(components, 'a'))
+            this.a = a;
+    }
+
+    // other color types
+    this(C)(C color) if(isColor!C)
+    {
+        // assignment operator performs format conversion
+        this = color;
+    }
+
+    // hex string initialiser
+    this(C)(const(C)[] hex) if(isSomeChar!C)
+    {
+        fromHex(hex);
+    }
+
+    // read hex strings in the standard forms: (#/$/0x)rgb/argb/rrggbb/aarrggbb
+    void fromHex(S)(const(S)[] hex) if(isSomeChar!S)
+    {
+        static ubyte val(ubyte c)
+        {
+            if(c >= '0' && c <= '9')
+                return cast(ubyte)(c - '0');
+            else if(c >= 'a' && c <= 'f')
+                return cast(ubyte)(c - 'a' + 10);
+            else if(c >= 'A' && c <= 'F')
+                return cast(ubyte)(c - 'A' + 10);
+            else
+                assert(false, "Invalid hex string");
+        }
+
+        if(hex.length > 0 && (hex[0] == '#' || hex[0] == '$'))
+            hex = hex[1..$];
+        else if(hex.length > 1 && (hex[0] == '0' && hex[1] == 'x'))
+            hex = hex[2..$];
+
+        if(hex.length == 3)
+        {
+            ubyte r = val(hex[0]);
+            ubyte g = val(hex[1]);
+            ubyte b = val(hex[2]);
+            this = RGB!("rgb", ubyte, colorSpace)(cast(ubyte)(r | (r << 4)), cast(ubyte)(g | (g << 4)), cast(ubyte)(b | (b << 4)));
+        }
+        if(hex.length == 4)
+        {
+            ubyte a = val(hex[0]);
+            ubyte r = val(hex[1]);
+            ubyte g = val(hex[2]);
+            ubyte b = val(hex[3]);
+            this = RGB!("rgba", ubyte, colorSpace)(cast(ubyte)(r | (r << 4)), cast(ubyte)(g | (g << 4)), cast(ubyte)(b | (b << 4)), cast(ubyte)(a | (a << 4)));
+        }
+        if(hex.length == 6)
+        {
+            ubyte r = cast(ubyte)(val(hex[0]) << 4) | val(hex[1]);
+            ubyte g = cast(ubyte)(val(hex[2]) << 4) | val(hex[3]);
+            ubyte b = cast(ubyte)(val(hex[4]) << 4) | val(hex[5]);
+            this = RGB!("rgb", ubyte, colorSpace)(r, g, b);
+        }
+        if(hex.length == 8)
+        {
+            ubyte a = cast(ubyte)(val(hex[0]) << 4) | val(hex[1]);
+            ubyte r = cast(ubyte)(val(hex[2]) << 4) | val(hex[3]);
+            ubyte g = cast(ubyte)(val(hex[4]) << 4) | val(hex[5]);
+            ubyte b = cast(ubyte)(val(hex[6]) << 4) | val(hex[7]);
+            this = RGB!("rgba", ubyte, colorSpace)(r, g, b, a);
+        }
+        else
+            assert(false, "Invalid hex string!");
+    }
+
+    // assignment
+    ref typeof(this) opAssign(C)(C rh) if(isColor!C)
+    {
+        auto arg = to!(typeof(this))(rh);
+        foreach(c; AllComponents)
+            mixin(ComponentExpression!("_ = arg._;", c, null));
```

Comment:
> Either, arithmetic works across colour types, or it doesn't.

Does it have to be that way? But if we can't have implicit conversion only when there's no data loss, then I'm leaning towards not having implicit conversion at all. Explicit conversion shouldn't be that much of a burden, and cases where implicit conversion is justified, as in your example above, will hopefully not occur that often.